## SF crime data analysis and modeling

### In this notebook, we will use Spark SQL for big data analysis on SF crime data. (https://data.sfgov.org/Public-Safety/Police-Department-Incident-Reports-Historical-2003/tmnf-yvry). 
1. The first part of this project is OLAP for scrime data analysis.
2. The second part is unsupervised learning for spatial data analysis

In [3]:
from csv import reader
from pyspark.sql import Row 
from pyspark.sql import SparkSession
from pyspark.sql.types import *
import pandas as pd
import numpy as np
import seaborn as sb
import matplotlib.pyplot as plt
import warnings

import os
os.environ["PYSPARK_PYTHON"] = "python3"


In [4]:
data_path = "dbfs:/laioffer/spark_hw1/data/sf_03_18.csv"
# use this file name later

In [5]:
#Get the DataFrame
from pyspark.sql import SparkSession
spark = SparkSession \
    .builder \
    .appName("crime analysis") \
    .config("spark.some.config.option", "some-value") \
    .getOrCreate()

df_opt1 = spark.read.format("csv").option("header", "true").load(data_path)
display(df_opt1)
df_opt1.createOrReplaceTempView("sf_crime")



IncidntNum Category Descript DayOfWeek Date Time PdDistrict Resolution Address X Y Location PdId SF Find Neighborhoods Current Police Districts Current Supervisor Districts Analysis Neighborhoods DELETE - Fire Prevention Districts DELETE - Police Districts DELETE - Supervisor Districts DELETE - Zip Codes DELETE - Neighborhoods DELETE - 2017 Fix It Zones Civic Center Harm Reduction Project Boundary Fix It Zones as of 2017-11-06 DELETE - HSOC Zones Fix It Zones as of 2018-02-07 CBD, BID and GBD Boundaries as of 2017 Areas of Vulnerability, 2016 Central Market/Tenderloin Boundary Central Market/Tenderloin Boundary Polygon - Updated HSOC Zones as of 2018-06-05 OWED Public Spaces 146196161 NON-CRIMINAL LOST PROPERTY Tuesday 09/23/2014 01:00 SOUTHERN NONE 800 Block of BRYANT ST -122.403404791479 37.775420706711 POINT (-122.403404791479 37.775420706711) 14619616171000 32 1 10 34 14 2 9 28853 34 null null null null null null 2 null null null null 150045675 ASSAULT BATTERY Thursday 01/15/2015 17:00 TARAVAL NONE 1800 Block of VICENTE ST -122.48560378101 37.7388214326705 POINT (-122.48560378101 37.7388214326705) 15004567504134 40 10 7 35 1 8 3 29491 35 null null null null null null 1 null null null null 140632022 SUSPICIOUS OCC INVESTIGATIVE DETENTION Wednesday 07/30/2014 09:32 BAYVIEW NONE 100 Block of GILLETTE AV -122.396535107224 37.7106603302503 POINT (-122.396535107224 37.7106603302503) 14063202264085 89 2 9 1 10 3 8 309 1 null null null null null null 1 null null null null 150383259 ASSAULT BATTERY Saturday 05/02/2015 23:10 BAYVIEW ARREST, BOOKED 2400 Block of PHELPS ST -122.400130573297 37.7300925390327 POINT (-122.400130573297 37.7300925390327) 15038325904134 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 040753980 OTHER OFFENSES RECKLESS DRIVING Friday 07/02/2004 13:43 BAYVIEW NONE I-280 / CESAR CHAVEZ ST -120.5 90 POINT (-120.5 90) 4075398065020 null null null null null null null null null null null null null null null null null null null null 040855122 SUICIDE SUICIDE BY JUMPING Tuesday 07/27/2004 15:19 SOUTHERN NONE 500 Block of I-80 -122.386667033903 37.7898821569191 POINT (-122.386667033903 37.7898821569191) 4085512260170 null null null null null null null 28856 null null null null null null null null null null null null 066085191 NON-CRIMINAL LOST PROPERTY Sunday 11/19/2006 17:45 BAYVIEW NONE 0 Block of GIANTS DR -122.38750147945 37.716878646429 POINT (-122.38750147945 37.716878646429) 6608519171000 88 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 050908404 VEHICLE THEFT STOLEN AUTOMOBILE Saturday 08/13/2005 17:00 TENDERLOIN NONE JENNINGS CT / INGALLS ST -120.5 90 POINT (-120.5 90) 5090840407021 null null null null null null null null null null null null null null null null null null null null 090768064 ARSON ARSON OF A VEHICLE Tuesday 07/28/2009 23:26 BAYVIEW NONE SELBY ST / OAKDALE AV -122.399686082806 37.739901780585 POINT (-122.399686082806 37.739901780585) 9076806426031 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 111027676 ASSAULT BATTERY Saturday 12/24/2011 07:00 SOUTHERN NONE 0 Block of DORE ST -122.412933062384 37.7739274524819 POINT (-122.412933062384 37.7739274524819) 11102767604134 32 1 10 34 8 2 9 28853 34 null 1 null 1 null null 2 null null 1 null 060508925 ASSAULT BATTERY Saturday 05/13/2006 11:26 NORTHERN NONE POLK ST / CEDAR ST -122.419834938618 37.7863580362743 POINT (-122.419834938618 37.7863580362743) 6050892504134 50 4 10 36 13 9 9 28858 36 null null null null null 10 2 null null null null 110318854 FRAUD FRAUDULENT CREDIT APPLICATION Thursday 04/14/2011 17:30 BAYVIEW NONE 100 Block of TOPEKA AV -122.398951090192 37.7338473162583 POINT (-122.398951090192 37.7338473162583) 11031885409340 87 2 9 1 10 3 8 58 1 null null null null null null 2 null null null null 131007921 SUSPICIOUS OCC SUSPICIOUS OCCURRENCE Thursday 11/28/2013 12:00 INGLESIDE NONE 1600 Block of DIAMOND ST -122.435709983124 37.7443688410754 POINT (-122.4357099831

#### (OLAP)Counts The Number of Crimes for Different Category.

In [7]:
q1_result = df_opt1.groupBy('category').count().orderBy('count', ascending=False)
display(q1_result)

category count LARCENY/THEFT 480448 OTHER OFFENSES 309358 NON-CRIMINAL 238323 ASSAULT 194694 VEHICLE THEFT 126602 DRUG/NARCOTIC 119628 VANDALISM 116059 WARRANTS 101379 BURGLARY 91543 SUSPICIOUS OCC 80444 MISSING PERSON 64961 ROBBERY 55867 FRAUD 41542 SECONDARY CODES 25831 FORGERY/COUNTERFEITING 23050 WEAPON LAWS 22234 TRESPASS 19449 PROSTITUTION 16701 STOLEN PROPERTY 11891 SEX OFFENSES, FORCIBLE 11742 DISORDERLY CONDUCT 10040 DRUNKENNESS 9826 RECOVERED VEHICLE 8716 DRIVING UNDER THE INFLUENCE 5672 KIDNAPPING 5346 RUNAWAY 4440 LIQUOR LAWS 4083 ARSON 3931 EMBEZZLEMENT 2988 LOITERING 2430 SUICIDE 1292 FAMILY OFFENSES 1183 BAD CHECKS 925 BRIBERY 813 EXTORTION 741 SEX OFFENSES, NON FORCIBLE 431 GAMBLING 348 PORNOGRAPHY/OBSCENE MAT 59 TREA 14

In [8]:
#Spark SQL based
crimeCategory = spark.sql("SELECT  category, COUNT(*) AS Count FROM sf_crime GROUP BY category ORDER BY Count DESC")
display(crimeCategory)

category Count LARCENY/THEFT 480448 OTHER OFFENSES 309358 NON-CRIMINAL 238323 ASSAULT 194694 VEHICLE THEFT 126602 DRUG/NARCOTIC 119628 VANDALISM 116059 WARRANTS 101379 BURGLARY 91543 SUSPICIOUS OCC 80444 MISSING PERSON 64961 ROBBERY 55867 FRAUD 41542 SECONDARY CODES 25831 FORGERY/COUNTERFEITING 23050 WEAPON LAWS 22234 TRESPASS 19449 PROSTITUTION 16701 STOLEN PROPERTY 11891 SEX OFFENSES, FORCIBLE 11742 DISORDERLY CONDUCT 10040 DRUNKENNESS 9826 RECOVERED VEHICLE 8716 DRIVING UNDER THE INFLUENCE 5672 KIDNAPPING 5346 RUNAWAY 4440 LIQUOR LAWS 4083 ARSON 3931 EMBEZZLEMENT 2988 LOITERING 2430 SUICIDE 1292 FAMILY OFFENSES 1183 BAD CHECKS 925 BRIBERY 813 EXTORTION 741 SEX OFFENSES, NON FORCIBLE 431 GAMBLING 348 PORNOGRAPHY/OBSCENE MAT 59 TREA 14

In [9]:
crimes_pd_df = crimeCategory.toPandas()

####(OLAP)Counts The Number of Crimes For Different District, and Visualize the Results

In [11]:

## use sql to compute the statistic result
crimeDistrict = spark.sql("SELECT PdDistrict, COUNT(*) AS Count FROM sf_crime GROUP BY 1 ORDER BY Count DESC")
display(crimeDistrict)

## visualize the results
crimes_pd_df_1 = crimeDistrict.toPandas()


PdDistrict Count SOUTHERN 399785 MISSION 300076 NORTHERN 272713 CENTRAL 226255 BAYVIEW 221000 INGLESIDE 194180 TENDERLOIN 191746 TARAVAL 166971 PARK 125479 RICHMOND 116818 null 1

#### (OLAP)Count The Number of Crimes Each "Sunday" at "SF downtown".

####Note:
SF downtown is defiend via the range of spatial location. For example, you can use a rectangle to define the SF downtown, or you can define a cicle with center as well. Thus, we write our own UDF function to filter data which are located inside certain spatial range.

In [14]:
#Using a udf to find all the crimes in down town SF.
from pyspark.sql.types import BooleanType

def within_geo_coordinates(x_in, y_in):
  if float(x_in) > -122.4313 and float(x_in) < -122.4213 and float(y_in) > 37.7540 and float(y_in) < 37.7740:
    return True
  else:
    return False
  
within_geo_coordinates_udf = udf(within_geo_coordinates, BooleanType())#register udf
df_opt1.withColumn('in_downTownSF', within_geo_coordinates_udf("X", "Y"))
df_q3=df_opt1.select("IncidntNum", "Category","Date", "DayOfWeek", within_geo_coordinates_udf("X", "Y").alias("in_downTownSF")).orderBy("in_downTownSF", ascending = False)
display(df_q3)



IncidntNum Category Date DayOfWeek in_downTownSF 041324328 ASSAULT 11/19/2004 Friday true 050025246 DRUG/NARCOTIC 01/07/2005 Friday true 051173814 OTHER OFFENSES 10/17/2005 Monday true 130437848 WARRANTS 05/27/2013 Monday true 050218601 VEHICLE THEFT 02/24/2005 Thursday true 100341984 LARCENY/THEFT 06/01/2009 Monday true 031276789 LARCENY/THEFT 10/30/2003 Thursday true 101003761 LARCENY/THEFT 10/28/2010 Thursday true 050587177 LARCENY/THEFT 05/25/2005 Wednesday true 130088140 NON-CRIMINAL 01/31/2013 Thursday true 030407761 VEHICLE THEFT 04/06/2003 Sunday true 030934758 WARRANTS 08/07/2003 Thursday true 106095315 LARCENY/THEFT 07/30/2010 Friday true 091022760 FORGERY/COUNTERFEITING 10/03/2009 Saturday true 110324265 OTHER OFFENSES 04/21/2011 Thursday true 030784412 LARCENY/THEFT 06/30/2003 Monday true 070613752 PROSTITUTION 06/18/2007 Monday true 060507143 RECOVERED VEHICLE 05/12/2006 Friday true 030150879 NON-CRIMINAL 02/05/2003 Wednesday true 050266503 LARCENY/THEFT 03/08/2005 Tuesday true 091164889 LARCENY/THEFT 11/11/2009 Wednesday true 060352584 OTHER OFFENSES 04/02/2006 Sunday true 080912970 DISORDERLY CONDUCT 08/07/2008 Thursday true 040113475 VEHICLE THEFT 01/27/2004 Tuesday true 100605382 ASSAULT 06/29/2010 Tuesday true 031492101 LARCENY/THEFT 12/24/2003 Wednesday true 110324265 WARRANTS 04/21/2011 Thursday true 070259221 WARRANTS 03/12/2007 Monday true 070215988 OTHER OFFENSES 03/01/2007 Thursday true 050832881 VANDALISM 07/31/2005 Sunday true 110545217 WARRANTS 07/06/2011 Wednesday true 120355103 VEHICLE THEFT 05/04/2012 Friday true 120416971 MISSING PERSON 05/25/2012 Friday true 130353426 TRESPASS 04/30/2013 Tuesday true 050029210 LARCENY/THEFT 01/08/2005 Saturday true 030418730 VEHICLE THEFT 04/08/2003 Tuesday true 100645009 OTHER OFFENSES 07/14/2010 Wednesday true 101013328 OTHER OFFENSES 10/31/2010 Sunday true 061045219 LARCENY/THEFT 09/27/2006 Wednesday true 080466660 ASSAULT 05/04/2008 Sunday true 100590545 VEHICLE THEFT 06/27/2010 Sunday true 120622883 VANDALISM 08/06/2012 Monday true 100118147 OTHER OFFENSES 02/05/2010 Friday true 101182967 VANDALISM 12/25/2010 Saturday true 096088808 LARCENY/THEFT 09/20/2009 Sunday true 030238942 MISSING PERSON 02/26/2003 Wednesday true 096040262 NON-CRIMINAL 05/01/2009 Friday true 081383920 OTHER OFFENSES 12/29/2008 Monday true 106164528 NON-CRIMINAL 12/25/2010 Saturday true 030220529 OTHER OFFENSES 02/22/2003 Saturday true 100450977 WARRANTS 05/15/2010 Saturday true 041203512 OTHER OFFENSES 10/21/2004 Thursday true 051031121 LARCENY/THEFT 09/13/2005 Tuesday true 100264853 ROBBERY 03/20/2010 Saturday true 100366396 LARCENY/THEFT 04/20/2010 Tuesday true 090968054 LARCENY/THEFT 09/19/2009 Saturday true 080087501 NON-CRIMINAL 01/25/2008 Friday true 060868525 VEHICLE THEFT 08/15/2006 Tuesday true 090461105 KIDNAPPING 05/03/2009 Sunday true 091186128 ASSAULT 11/17/2009 Tuesday true 100684223 DRUG/NARCOTIC 07/26/2010 Monday true 050310207 OTHER OFFENSES 03/20/2005 Sunday true 040244741 DRUG/NARCOTIC 03/01/2004 Monday true 060558356 OTHER OFFENSES 05/26/2006 Friday true 070401927 NON-CRIMINAL 04/19/2007 Thursday true 071036703 ROBBERY 10/09/2007 Tuesday true 110314517 VEHICLE THEFT 04/17/2011 Sunday true 030372409 VEHICLE THEFT 03/30/2003 Sunday true 120121089 BURGLARY 01/15/2012 Sunday true 040263161 FRAUD 03/05/2004 Friday true 051314507 NON-CRIMINAL 11/21/2005 Monday true 090645280 FORGERY/COUNTERFEITING 06/24/2009 Wednesday true 136211484 LARCENY/THEFT 11/08/2013 Friday true 116139135 LARCENY/THEFT 10/26/2011 Wednesday true 070362313 VEHICLE THEFT 02/09/2007 Friday true 070708187 NON-CRIMINAL 07/14/2007 Saturday true 050879742 DRUG/NARCOTIC 08/06/2005 Saturday true 040299144 SUSPICIOUS OCC 03/07/2004 Sunday true 070698942 OTHER OFFENSES 07/11/2007 Wednesday true 081070183 OTHER OFFENSES 10/06/2008 Monday true 120026928 SECONDARY CODES 01/09/2012 Monday true 090032370 SECONDARY CODES 01/10/2009 Saturday true 116130672 LARCENY/THEFT 10/16/2011 Sunday true 120050030 FRAUD 01/03/2

In [15]:
#count crime number of each sunday

df_q3.createOrReplaceTempView("sf_crime_q3")
crimeDowntown = spark.sql("SELECT Date, Count(*) AS Count FROM sf_crime_q3 WHERE DayOfWeek = 'Sunday' GROUP By Date Order By Date ASC")
display(crimeDowntown)

Date Count 01/01/2006 621 01/01/2012 530 01/01/2017 485 01/02/2005 304 01/02/2011 305 01/03/2010 356 01/03/2016 316 01/04/2004 351 01/04/2009 350 01/04/2015 335 01/05/2003 377 01/05/2014 320 01/06/2008 375 01/06/2013 342 01/07/2007 397 01/07/2018 406 01/08/2006 348 01/08/2012 347 01/08/2017 368 01/09/2005 361 01/09/2011 339 01/10/2010 358 01/10/2016 402 01/11/2004 339 01/11/2009 378 01/11/2015 368 01/12/2003 324 01/12/2014 369 01/13/2008 378 01/13/2013 340 01/14/2007 302 01/14/2018 380 01/15/2006 349 01/15/2012 348 01/15/2017 419 01/16/2005 365 01/16/2011 356 01/17/2010 318 01/17/2016 387 01/18/2004 360 01/18/2009 358 01/18/2015 401 01/19/2003 305 01/19/2014 390 01/20/2008 378 01/20/2013 416 01/21/2007 340 01/21/2018 343 01/22/2006 358 01/22/2012 340 01/22/2017 418 01/23/2005 358 01/23/2011 334 01/24/2010 357 01/24/2016 400 01/25/2004 405 01/25/2009 343 01/25/2015 365 01/26/2003 341 01/26/2014 405 01/27/2008 316 01/27/2013 386 01/28/2007 342 01/28/2018 352 01/29/2006 341 01/29/2012 370 01/29/2017 424 01/30/2005 375 01/30/2011 313 01/31/2010 297 01/31/2016 402 02/01/2004 370 02/01/2009 417 02/01/2015 466 02/02/2003 355 02/02/2014 361 02/03/2008 307 02/03/2013 397 02/04/2007 341 02/04/2018 301 02/05/2006 326 02/05/2012 359 02/05/2017 386 02/06/2005 307 02/06/2011 349 02/07/2010 312 02/07/2016 460 02/08/2004 339 02/08/2009 297 02/08/2015 394 02/09/2003 360 02/09/2014 365 02/10/2008 376 02/10/2013 393 02/11/2007 351 02/11/2018 323 02/12/2006 385 02/12/2012 394 02/12/2017 369 02/13/2005 323 02/13/2011 337 02/14/2010 389 02/14/2016 365 02/15/2004 413 02/15/2009 328 02/15/2015 416 02/16/2003 332 02/16/2014 386 02/17/2008 348 02/17/2013 444 02/18/2007 329 02/18/2018 335 02/19/2006 344 02/19/2012 345 02/19/2017 421 02/20/2005 331 02/20/2011 329 02/21/2010 280 02/21/2016 424 02/22/2004 354 02/22/2009 318 02/22/2015 445 02/23/2003 366 02/23/2014 367 02/24/2008 370 02/24/2013 379 02/25/2007 285 02/25/2018 369 02/26/2006 341 02/26/2012 324 02/26/2017 408 02/27/2005 288 02/27/2011 307 02/28/2010 358 02/28/2016 382 02/29/2004 389 03/01/2009 455 03/01/2015 475 03/02/2003 358 03/02/2008 393 03/02/2014 385 03/03/2013 381 03/04/2007 371 03/04/2012 363 03/04/2018 356 03/05/2006 286 03/05/2017 383 03/06/2005 320 03/06/2011 298 03/06/2016 368 03/07/2004 360 03/07/2010 358 03/08/2009 356 03/08/2015 412 03/09/2003 359 03/09/2008 377 03/09/2014 378 03/10/2013 345 03/11/2007 338 03/11/2012 392 03/11/2018 293 03/12/2006 293 03/12/2017 406 03/13/2005 367 03/13/2011 349 03/13/2016 309 03/14/2004 297 03/14/2010 367 03/15/2009 363 03/15/2015 482 03/16/2003 363 03/16/2008 364 03/16/2014 355 03/17/2013 428 03/18/2007 352 03/18/2012 366 03/18/2018 333 03/19/2006 364 03/19/2017 365 03/20/2005 370 03/20/2011 339 03/20/2016 360 03/21/2004 372 03/21/2010 318 03/22/2009 369 03/22/2015 430 03/23/2003 323 03/23/2008 366 03/23/2014 402 03/24/2013 414 03/25/2007 398 03/25/2012 362 03/25/2018 349 03/26/2006 365 03/26/2017 456 03/27/2005 321 03/27/2011 318 03/27/2016 387 03/28/2004 405 03/28/2010 303 03/29/2009 450 03/29/2015 416 03/30/2003 396 03/30/2008 326 03/30/2014 338 03/31/2013 371 04/01/2007 454 04/01/2012 393 04/01/2018 327 04/02/2006 363 04/02/2017 375 04/03/2005 326 04/03/2011 350 04/03/2016 384 04/04/2004 391 04/04/2010 232 04/05/2009 380 04/05/2015 384 04/06/2003 343 04/06/2008 354 04/06/2014 409 04/07/2013 402 04/08/2007 292 04/08/2012 327 04/08/2018 299 04/09/2006 352 04/09/2017 335 04/10/2005 313 04/10/2011 319 04/10/2016 338 04/11/2004 331 04/11/2010 252 04/12/2009 363 04/12/2015 405 04/13/2003 389 04/13/2008 368 04/13/2014 353 04/14/2013 382 04/15/2007 351 04/15/2012 364 04/15/2018 301 04/16/2006 322 04/16/2017 389 04/17/2005 320 04/17/2011 324 04/17/2016 374 04/18/2004 389 04/18/2010 274 04/19/2009 381 04/19/2015 462 04/20/2003 369 04/20/2008 430 04/20/2014 365 04/21/2013 407 04/22/2007 366 04/22/2012 326 04/22/2018 302 04/23/2006 371 04/23/2017 407 04/24/2005 363 04/24/2011 289 04/24/2016 382 04/25/2004 430 04/25/2010 352 04/26/2009 370 04/26/2015 404

#### (OLAP)Analysis The Number of Crime in Each Month of 2015, 2016, 2017, 2018.

In [17]:
#%sql select substring(Date,7) as Year, substring(Date,1,2) as Month, count(*)  from sf_crime
#group by Year, Month
#having Year Between 2015 and 2018
#order by Year, Month
df_opt1.createOrReplaceTempView("sf_crime")
temp = spark.sql("select substring(Date,7) as Year, substring(Date,1,2) as Month, count(*)  from sf_crime group by Year, Month having Year Between 2015 and 2018 order by Year, Month")
display(temp)


Year Month count(1) 2015 01 13606 2015 02 12329 2015 03 13929 2015 04 12959 2015 05 13729 2015 06 13304 2015 07 13365 2015 08 13730 2015 09 12896 2015 10 13147 2015 11 12091 2015 12 11441 2016 01 12967 2016 02 12106 2016 03 12380 2016 04 12328 2016 05 12732 2016 06 12094 2016 07 12191 2016 08 12471 2016 09 12499 2016 10 13388 2016 11 12720 2016 12 13007 2017 01 13084 2017 02 12192 2017 03 13711 2017 04 12941 2017 05 13267 2017 06 12605 2017 07 13171 2017 08 12872 2017 09 12684 2017 10 13355 2017 11 12347 2017 12 12544 2018 01 12031 2018 02 9947 2018 03 10740 2018 04 10306 2018 05 3644

In [18]:
df_opt1.createOrReplaceTempView("sf_crime")
temp = spark.sql("select substring(Date,7) as Year, substring(Date,1,2) as Month, count(*)  from sf_crime group by Year, Month having Year Between 2015 and 2017 order by Year, Month")
display(temp)

Year Month count(1) 2015 01 13606 2015 02 12329 2015 03 13929 2015 04 12959 2015 05 13729 2015 06 13304 2015 07 13365 2015 08 13730 2015 09 12896 2015 10 13147 2015 11 12091 2015 12 11441 2016 01 12967 2016 02 12106 2016 03 12380 2016 04 12328 2016 05 12732 2016 06 12094 2016 07 12191 2016 08 12471 2016 09 12499 2016 10 13388 2016 11 12720 2016 12 13007 2017 01 13084 2017 02 12192 2017 03 13711 2017 04 12941 2017 05 13267 2017 06 12605 2017 07 13171 2017 08 12872 2017 09 12684 2017 10 13355 2017 11 12347 2017 12 12544

####Note (Insight)
1. Because 2018 has a significant different pattern from the previous years and missed almost half year's data, we viewed it as an exception and exclude it from the analysis. 
2. The graph shows a cycling pattern with higher crime rate months immediately follow by lower crime rate months. This fluctuation can be caused by the intensity of police enforcement follows the crime rate. When the crime rate is high, the police enforcement intensifies. Then the crime rate drops. The enforcement relaxes after the drops which then follow by a higher crime rate. To smooth the crime curve over months, law enforcement can place the resource more evenly in the future. And police can be more alert and prepared for the high crime month even in a low crime month
3. Based on the data from 2015, 2016 and 2017, San Francisco has similar monthly crime pattern in different years. The crime rate always decreases significantly from January to February and bounce back from February to March. The number of crimes also falls from October to the end of the year. Due to the lower crime rate in each year's February and October, event holders and business owners may consider scheduling their events and visits in these two months with lower crime rate. 
4. Based on the data, the January and March usually have a higher crime rate relatively. It can be caused by the concentration of holidays in these two months. For example, in each year, SF will hold a parade for St.Patrick day in March. Due to the concentration of holidays in January and March, the number of tourists increases which increase the crime rate. 
5. The charts also show a decrease of crime rate after October. From some crime study, temperature can affect people's aggressive behavior. The lower temperature in winter days in SF can be the reason behind the lower crime rate

#### (OLAP)Analysis The Number of Crime w.r.t The Hour in December 15 Each Year

In [21]:
#Change the date format of the date column
from pyspark.sql.functions import *
df_update = df_opt1.withColumn("Date", to_date(col("Date"), "MM/dd/yyyy")) ##change datetype from string to date
df_update.createOrReplaceTempView("sf_crime")

temp = spark.sql("Select Date, hour(Time) As Hour, Count(*) as Count From sf_crime Group By Date, Hour Having Year(Date) between 2015 and 2017 and Month(Date)=12 and DayOfMonth(Date)=15 Order By 1,2")
display(temp)

Date Hour Count 2015-12-15 0 15 2015-12-15 1 6 2015-12-15 2 5 2015-12-15 3 4 2015-12-15 4 10 2015-12-15 5 3 2015-12-15 6 4 2015-12-15 7 8 2015-12-15 8 12 2015-12-15 9 10 2015-12-15 10 26 2015-12-15 11 15 2015-12-15 12 35 2015-12-15 13 13 2015-12-15 14 27 2015-12-15 15 14 2015-12-15 16 31 2015-12-15 17 18 2015-12-15 18 18 2015-12-15 19 26 2015-12-15 20 13 2015-12-15 21 13 2015-12-15 22 13 2015-12-15 23 10 2016-12-15 0 22 2016-12-15 1 10 2016-12-15 2 12 2016-12-15 3 9 2016-12-15 4 1 2016-12-15 5 6 2016-12-15 6 5 2016-12-15 7 3 2016-12-15 8 24 2016-12-15 9 23 2016-12-15 10 13 2016-12-15 11 17 2016-12-15 12 29 2016-12-15 13 18 2016-12-15 14 19 2016-12-15 15 13 2016-12-15 16 17 2016-12-15 17 20 2016-12-15 18 38 2016-12-15 19 31 2016-12-15 20 27 2016-12-15 21 16 2016-12-15 22 22 2016-12-15 23 16 2017-12-15 0 30 2017-12-15 1 12 2017-12-15 2 8 2017-12-15 3 4 2017-12-15 4 9 2017-12-15 5 2 2017-12-15 6 6 2017-12-15 7 11 2017-12-15 8 26 2017-12-15 9 14 2017-12-15 10 27 2017-12-15 11 26 2017-12-15 12 24 2017-12-15 13 10 2017-12-15 14 12 2017-12-15 15 27 2017-12-15 16 31 2017-12-15 17 28 2017-12-15 18 29 2017-12-15 19 31 2017-12-15 20 19 2017-12-15 21 20 2017-12-15 22 37 2017-12-15 23 28

1. We first calculated the crime numbers in every hour for 12/15/2015, 12/15/2016 and 12/15/2017. In order to see the pattern more clearly, we stacked the lines. From the line chart, the crime number decreased from mid night and hit the bottom before sunrise. After sunrise, the crime rate begins to increase. 
2.  In order to know if this pattern is only a coincident for December 15, we then accumulated all the crime numbers each hours for all the dates and have the below line chart.

In [23]:
temp = spark.sql("Select Date, hour(Time) As Hour, Count(*) as Count From sf_crime Group By Date, Hour Order By 1,2")
display(temp)

Date Hour Count 2003-01-01 0 103 2003-01-01 1 31 2003-01-01 2 38 2003-01-01 3 17 2003-01-01 4 5 2003-01-01 5 7 2003-01-01 6 10 2003-01-01 7 21 2003-01-01 8 18 2003-01-01 9 25 2003-01-01 10 17 2003-01-01 11 7 2003-01-01 12 57 2003-01-01 13 11 2003-01-01 14 17 2003-01-01 15 28 2003-01-01 16 31 2003-01-01 17 25 2003-01-01 18 15 2003-01-01 19 28 2003-01-01 20 24 2003-01-01 21 27 2003-01-01 22 33 2003-01-01 23 27 2003-01-02 0 25 2003-01-02 1 6 2003-01-02 2 8 2003-01-02 3 5 2003-01-02 4 5 2003-01-02 5 2 2003-01-02 6 8 2003-01-02 7 9 2003-01-02 8 10 2003-01-02 9 8 2003-01-02 10 15 2003-01-02 11 15 2003-01-02 12 26 2003-01-02 13 16 2003-01-02 14 23 2003-01-02 15 22 2003-01-02 16 29 2003-01-02 17 26 2003-01-02 18 32 2003-01-02 19 17 2003-01-02 20 32 2003-01-02 21 29 2003-01-02 22 20 2003-01-02 23 23 2003-01-03 0 17 2003-01-03 1 8 2003-01-03 2 15 2003-01-03 3 3 2003-01-03 4 3 2003-01-03 5 5 2003-01-03 6 7 2003-01-03 7 23 2003-01-03 8 24 2003-01-03 9 19 2003-01-03 10 27 2003-01-03 11 23 2003-01-03 12 24 2003-01-03 13 14 2003-01-03 14 16 2003-01-03 15 21 2003-01-03 16 18 2003-01-03 17 21 2003-01-03 18 36 2003-01-03 19 23 2003-01-03 20 23 2003-01-03 21 28 2003-01-03 22 26 2003-01-03 23 16 2003-01-04 0 17 2003-01-04 1 21 2003-01-04 2 11 2003-01-04 3 5 2003-01-04 4 2 2003-01-04 5 2 2003-01-04 6 9 2003-01-04 7 8 2003-01-04 8 14 2003-01-04 9 11 2003-01-04 10 14 2003-01-04 11 25 2003-01-04 12 15 2003-01-04 13 18 2003-01-04 14 13 2003-01-04 15 16 2003-01-04 16 18 2003-01-04 17 16 2003-01-04 18 30 2003-01-04 19 10 2003-01-04 20 8 2003-01-04 21 17 2003-01-04 22 29 2003-01-04 23 18 2003-01-05 0 17 2003-01-05 1 21 2003-01-05 2 12 2003-01-05 3 11 2003-01-05 4 4 2003-01-05 6 5 2003-01-05 8 18 2003-01-05 9 14 2003-01-05 10 20 2003-01-05 11 21 2003-01-05 12 27 2003-01-05 13 22 2003-01-05 14 35 2003-01-05 15 14 2003-01-05 16 15 2003-01-05 17 21 2003-01-05 18 19 2003-01-05 19 22 2003-01-05 20 8 2003-01-05 21 18 2003-01-05 22 15 2003-01-05 23 18 2003-01-06 0 14 2003-01-06 1 6 2003-01-06 2 12 2003-01-06 3 3 2003-01-06 4 1 2003-01-06 5 10 2003-01-06 6 5 2003-01-06 7 10 2003-01-06 8 24 2003-01-06 9 16 2003-01-06 10 25 2003-01-06 11 29 2003-01-06 12 13 2003-01-06 13 16 2003-01-06 14 21 2003-01-06 15 15 2003-01-06 16 17 2003-01-06 17 29 2003-01-06 18 37 2003-01-06 19 22 2003-01-06 20 20 2003-01-06 21 14 2003-01-06 22 21 2003-01-06 23 20 2003-01-07 0 17 2003-01-07 1 8 2003-01-07 2 9 2003-01-07 3 6 2003-01-07 4 3 2003-01-07 5 2 2003-01-07 6 5 2003-01-07 7 16 2003-01-07 8 19 2003-01-07 9 18 2003-01-07 10 19 2003-01-07 11 28 2003-01-07 12 24 2003-01-07 13 28 2003-01-07 14 21 2003-01-07 15 16 2003-01-07 16 25 2003-01-07 17 18 2003-01-07 18 33 2003-01-07 19 21 2003-01-07 20 26 2003-01-07 21 26 2003-01-07 22 19 2003-01-07 23 11 2003-01-08 0 20 2003-01-08 1 12 2003-01-08 2 10 2003-01-08 3 5 2003-01-08 4 11 2003-01-08 5 16 2003-01-08 6 5 2003-01-08 7 21 2003-01-08 8 22 2003-01-08 9 32 2003-01-08 10 31 2003-01-08 11 28 2003-01-08 12 34 2003-01-08 13 15 2003-01-08 14 23 2003-01-08 15 25 2003-01-08 16 28 2003-01-08 17 28 2003-01-08 18 24 2003-01-08 19 51 2003-01-08 20 26 2003-01-08 21 18 2003-01-08 22 25 2003-01-08 23 18 2003-01-09 0 21 2003-01-09 1 9 2003-01-09 2 10 2003-01-09 3 5 2003-01-09 4 7 2003-01-09 5 3 2003-01-09 6 9 2003-01-09 7 11 2003-01-09 8 29 2003-01-09 9 23 2003-01-09 10 16 2003-01-09 11 20 2003-01-09 12 25 2003-01-09 13 24 2003-01-09 14 25 2003-01-09 15 16 2003-01-09 16 19 2003-01-09 17 15 2003-01-09 18 23 2003-01-09 19 15 2003-01-09 20 18 2003-01-09 21 16 2003-01-09 22 28 2003-01-09 23 30 2003-01-10 0 12 2003-01-10 1 7 2003-01-10 2 6 2003-01-10 3 6 2003-01-10 4 1 2003-01-10 5 7 2003-01-10 6 3 2003-01-10 7 9 2003-01-10 8 15 2003-01-10 9 19 2003-01-10 10 18 2003-01-10 11 13 2003-01-10 12 29 2003-01-10 13 24 2003-01-10 14 19 2003-01-10 15 25 2003-01-10 16 16 2003-01-10 17 19 2003-01-10 18 27 2003-01-10 19 26 2003-01-10 20 13 2003-01-10 21 27 2003-01-10 22 26 2003-01-10 23 32 2003-01-11 0 24 2003-01-11 1 12 2003-01-11 2 11 2003-01-11 3 5 2003-01-11 4 6 2003-0

#### Note(Insight)
1. The graph shows a U shape from mid night to noon. 
2. Based on our crime data, san Francisco has lower crime numbers in the morning especially before the sunrise compared to the rest of the day. Morning is safer than afternoon. We will suggest traveler schedule their visit in the morning around 5am to 8am of a day.

#### (OLAP)Find out The Top-3 Danger District  
1. find out the crime event w.r.t category and time (hour) from the results. 
2. give your advice to distribute the police based on the analysis results.

In [26]:
#Find out the top-3 danger district
crimeDistrictTop3 = spark.sql("SELECT PdDistrict, COUNT(*) AS Count FROM sf_crime GROUP BY 1 ORDER BY Count DESC LIMIT 3")
display(crimeDistrictTop3)

PdDistrict Count SOUTHERN 399785 MISSION 300076 NORTHERN 272713

Based on the number of crime in each district, we found that the three most dangerours district in SF are Southern, Mission and Northern.

In [28]:
#find out the crime event w.r.t category and time (hour) from the result of step 1
#First of all, we pull out  the crime event data w.r.t category and time for Southern District.
crimeDistrictSouthern = spark.sql("SELECT PdDistrict, Category, hour(Time) As Hour, COUNT(*) AS Count FROM sf_crime GROUP BY 1,2,3 having PdDistrict = 'SOUTHERN' ORDER BY 1, 2,3 ASC ")
display(crimeDistrictSouthern)

PdDistrict Category Hour Count SOUTHERN ARSON 0 31 SOUTHERN ARSON 1 27 SOUTHERN ARSON 2 25 SOUTHERN ARSON 3 21 SOUTHERN ARSON 4 20 SOUTHERN ARSON 5 20 SOUTHERN ARSON 6 14 SOUTHERN ARSON 7 9 SOUTHERN ARSON 8 17 SOUTHERN ARSON 9 7 SOUTHERN ARSON 10 17 SOUTHERN ARSON 11 18 SOUTHERN ARSON 12 16 SOUTHERN ARSON 13 16 SOUTHERN ARSON 14 20 SOUTHERN ARSON 15 20 SOUTHERN ARSON 16 10 SOUTHERN ARSON 17 21 SOUTHERN ARSON 18 22 SOUTHERN ARSON 19 33 SOUTHERN ARSON 20 30 SOUTHERN ARSON 21 16 SOUTHERN ARSON 22 27 SOUTHERN ARSON 23 27 SOUTHERN ASSAULT 0 1714 SOUTHERN ASSAULT 1 1506 SOUTHERN ASSAULT 2 1466 SOUTHERN ASSAULT 3 697 SOUTHERN ASSAULT 4 400 SOUTHERN ASSAULT 5 342 SOUTHERN ASSAULT 6 502 SOUTHERN ASSAULT 7 724 SOUTHERN ASSAULT 8 1059 SOUTHERN ASSAULT 9 1163 SOUTHERN ASSAULT 10 1334 SOUTHERN ASSAULT 11 1387 SOUTHERN ASSAULT 12 1660 SOUTHERN ASSAULT 13 1514 SOUTHERN ASSAULT 14 1597 SOUTHERN ASSAULT 15 1700 SOUTHERN ASSAULT 16 1762 SOUTHERN ASSAULT 17 1733 SOUTHERN ASSAULT 18 1636 SOUTHERN ASSAULT 19 1575 SOUTHERN ASSAULT 20 1518 SOUTHERN ASSAULT 21 1481 SOUTHERN ASSAULT 22 1446 SOUTHERN ASSAULT 23 1398 SOUTHERN BAD CHECKS 0 35 SOUTHERN BAD CHECKS 1 1 SOUTHERN BAD CHECKS 2 1 SOUTHERN BAD CHECKS 4 2 SOUTHERN BAD CHECKS 5 1 SOUTHERN BAD CHECKS 6 1 SOUTHERN BAD CHECKS 7 1 SOUTHERN BAD CHECKS 8 7 SOUTHERN BAD CHECKS 9 11 SOUTHERN BAD CHECKS 10 9 SOUTHERN BAD CHECKS 11 6 SOUTHERN BAD CHECKS 12 27 SOUTHERN BAD CHECKS 13 8 SOUTHERN BAD CHECKS 14 12 SOUTHERN BAD CHECKS 15 14 SOUTHERN BAD CHECKS 16 7 SOUTHERN BAD CHECKS 17 12 SOUTHERN BAD CHECKS 18 6 SOUTHERN BAD CHECKS 19 5 SOUTHERN BAD CHECKS 20 5 SOUTHERN BAD CHECKS 21 2 SOUTHERN BAD CHECKS 22 2 SOUTHERN BAD CHECKS 23 1 SOUTHERN BRIBERY 0 8 SOUTHERN BRIBERY 1 3 SOUTHERN BRIBERY 2 3 SOUTHERN BRIBERY 3 1 SOUTHERN BRIBERY 4 2 SOUTHERN BRIBERY 6 1 SOUTHERN BRIBERY 7 1 SOUTHERN BRIBERY 8 3 SOUTHERN BRIBERY 9 3 SOUTHERN BRIBERY 10 9 SOUTHERN BRIBERY 11 5 SOUTHERN BRIBERY 12 7 SOUTHERN BRIBERY 13 4 SOUTHERN BRIBERY 14 5 SOUTHERN BRIBERY 15 4 SOUTHERN BRIBERY 16 5 SOUTHERN BRIBERY 17 9 SOUTHERN BRIBERY 18 5 SOUTHERN BRIBERY 19 5 SOUTHERN BRIBERY 20 4 SOUTHERN BRIBERY 21 5 SOUTHERN BRIBERY 22 8 SOUTHERN BRIBERY 23 8 SOUTHERN BURGLARY 0 486 SOUTHERN BURGLARY 1 359 SOUTHERN BURGLARY 2 400 SOUTHERN BURGLARY 3 417 SOUTHERN BURGLARY 4 361 SOUTHERN BURGLARY 5 328 SOUTHERN BURGLARY 6 294 SOUTHERN BURGLARY 7 243 SOUTHERN BURGLARY 8 369 SOUTHERN BURGLARY 9 359 SOUTHERN BURGLARY 10 390 SOUTHERN BURGLARY 11 417 SOUTHERN BURGLARY 12 613 SOUTHERN BURGLARY 13 432 SOUTHERN BURGLARY 14 545 SOUTHERN BURGLARY 15 669 SOUTHERN BURGLARY 16 710 SOUTHERN BURGLARY 17 1127 SOUTHERN BURGLARY 18 1060 SOUTHERN BURGLARY 19 820 SOUTHERN BURGLARY 20 657 SOUTHERN BURGLARY 21 523 SOUTHERN BURGLARY 22 504 SOUTHERN BURGLARY 23 468 SOUTHERN DISORDERLY CONDUCT 0 37 SOUTHERN DISORDERLY CONDUCT 1 24 SOUTHERN DISORDERLY CONDUCT 2 25 SOUTHERN DISORDERLY CONDUCT 3 18 SOUTHERN DISORDERLY CONDUCT 4 9 SOUTHERN DISORDERLY CONDUCT 5 37 SOUTHERN DISORDERLY CONDUCT 6 108 SOUTHERN DISORDERLY CONDUCT 7 160 SOUTHERN DISORDERLY CONDUCT 8 134 SOUTHERN DISORDERLY CONDUCT 9 110 SOUTHERN DISORDERLY CONDUCT 10 80 SOUTHERN DISORDERLY CONDUCT 11 57 SOUTHERN DISORDERLY CONDUCT 12 56 SOUTHERN DISORDERLY CONDUCT 13 56 SOUTHERN DISORDERLY CONDUCT 14 56 SOUTHERN DISORDERLY CONDUCT 15 34 SOUTHERN DISORDERLY CONDUCT 16 54 SOUTHERN DISORDERLY CONDUCT 17 40 SOUTHERN DISORDERLY CONDUCT 18 53 SOUTHERN DISORDERLY CONDUCT 19 28 SOUTHERN DISORDERLY CONDUCT 20 28 SOUTHERN DISORDERLY CONDUCT 21 40 SOUTHERN DISORDERLY CONDUCT 22 57 SOUTHERN DISORDERLY CONDUCT 23 34 SOUTHERN DRIVING UNDER THE INFLUENCE 0 87 SOUTHERN DRIVING UNDER THE INFLUENCE 1 101 SOUTHERN DRIVING UNDER THE INFLUENCE 2 84 SOUTHERN DRIVING UNDER THE INFLUENCE 3 41 SOUTHERN DRIVING UNDER THE INFLUENCE 4 12 SOUTHERN DRIVING UNDER THE INFLUENCE 5 7 SOUTHERN DRIVING UNDER THE INFLUENCE 6 15 SOUTHERN DRIVING UNDER THE INFLUENCE 7 5 SOUTHERN DRIVING UNDER THE INFLUENCE 8 4 SOUTHERN DRIVING UNDER THE INFLUENCE 9 12 SOUT

#### Note(Insight)
1. Crimes at all categories tend to follow the rule we previously discovered for SF city, which is the crime number decrease from mid night and increase after sunrise. That means Southern District is safer at night before sun rise and in the morning before noon.
2. Based on the line chart depicted by calculating the crime number distribution during a day respect to different categories, we found that the most common crime in Southern District is theft and these type of crime events concentrated at night. The crime in this category peaks between 18 and 19 o'clock.
3. The second and the third most serious crimes in this district are Assault and Drug. The Assault category has a smooth distribution which means the crime happened evenly at each hour during the day. 
4. The Drug category crime increase during the day and decrease after sunset.

In [30]:
#Secondly, we pull out  the crime event data w.r.t category and time for Mission District.
crimeDistrictMission = spark.sql("SELECT PdDistrict, Category, hour(Time) As Hour, COUNT(*) AS Count FROM sf_crime GROUP BY 1,2,3 having PdDistrict = 'MISSION' ORDER BY 1, 2,3 ASC ")
display(crimeDistrictMission)

PdDistrict Category Hour Count MISSION ARSON 0 35 MISSION ARSON 1 23 MISSION ARSON 2 28 MISSION ARSON 3 36 MISSION ARSON 4 27 MISSION ARSON 5 26 MISSION ARSON 6 12 MISSION ARSON 7 9 MISSION ARSON 8 7 MISSION ARSON 9 14 MISSION ARSON 10 12 MISSION ARSON 11 5 MISSION ARSON 12 14 MISSION ARSON 13 12 MISSION ARSON 14 14 MISSION ARSON 15 17 MISSION ARSON 16 16 MISSION ARSON 17 12 MISSION ARSON 18 19 MISSION ARSON 19 13 MISSION ARSON 20 24 MISSION ARSON 21 24 MISSION ARSON 22 16 MISSION ARSON 23 19 MISSION ASSAULT 0 1801 MISSION ASSAULT 1 1645 MISSION ASSAULT 2 1391 MISSION ASSAULT 3 644 MISSION ASSAULT 4 456 MISSION ASSAULT 5 317 MISSION ASSAULT 6 385 MISSION ASSAULT 7 557 MISSION ASSAULT 8 837 MISSION ASSAULT 9 988 MISSION ASSAULT 10 1036 MISSION ASSAULT 11 1080 MISSION ASSAULT 12 1335 MISSION ASSAULT 13 1288 MISSION ASSAULT 14 1339 MISSION ASSAULT 15 1460 MISSION ASSAULT 16 1479 MISSION ASSAULT 17 1545 MISSION ASSAULT 18 1535 MISSION ASSAULT 19 1477 MISSION ASSAULT 20 1532 MISSION ASSAULT 21 1507 MISSION ASSAULT 22 1620 MISSION ASSAULT 23 1537 MISSION BAD CHECKS 0 13 MISSION BAD CHECKS 1 1 MISSION BAD CHECKS 2 1 MISSION BAD CHECKS 6 1 MISSION BAD CHECKS 7 1 MISSION BAD CHECKS 8 12 MISSION BAD CHECKS 9 5 MISSION BAD CHECKS 10 4 MISSION BAD CHECKS 11 4 MISSION BAD CHECKS 12 19 MISSION BAD CHECKS 13 2 MISSION BAD CHECKS 14 5 MISSION BAD CHECKS 15 6 MISSION BAD CHECKS 16 3 MISSION BAD CHECKS 17 8 MISSION BAD CHECKS 18 1 MISSION BAD CHECKS 19 4 MISSION BAD CHECKS 20 1 MISSION BAD CHECKS 23 1 MISSION BRIBERY 0 9 MISSION BRIBERY 1 5 MISSION BRIBERY 2 7 MISSION BRIBERY 3 5 MISSION BRIBERY 4 2 MISSION BRIBERY 5 2 MISSION BRIBERY 6 1 MISSION BRIBERY 7 3 MISSION BRIBERY 8 2 MISSION BRIBERY 9 6 MISSION BRIBERY 10 8 MISSION BRIBERY 11 8 MISSION BRIBERY 12 7 MISSION BRIBERY 13 6 MISSION BRIBERY 14 8 MISSION BRIBERY 15 5 MISSION BRIBERY 16 7 MISSION BRIBERY 17 13 MISSION BRIBERY 18 6 MISSION BRIBERY 19 6 MISSION BRIBERY 20 9 MISSION BRIBERY 21 12 MISSION BRIBERY 22 9 MISSION BRIBERY 23 4 MISSION BURGLARY 0 482 MISSION BURGLARY 1 302 MISSION BURGLARY 2 315 MISSION BURGLARY 3 371 MISSION BURGLARY 4 310 MISSION BURGLARY 5 256 MISSION BURGLARY 6 223 MISSION BURGLARY 7 270 MISSION BURGLARY 8 495 MISSION BURGLARY 9 401 MISSION BURGLARY 10 327 MISSION BURGLARY 11 321 MISSION BURGLARY 12 419 MISSION BURGLARY 13 281 MISSION BURGLARY 14 337 MISSION BURGLARY 15 355 MISSION BURGLARY 16 492 MISSION BURGLARY 17 610 MISSION BURGLARY 18 640 MISSION BURGLARY 19 539 MISSION BURGLARY 20 498 MISSION BURGLARY 21 431 MISSION BURGLARY 22 511 MISSION BURGLARY 23 442 MISSION DISORDERLY CONDUCT 0 191 MISSION DISORDERLY CONDUCT 1 131 MISSION DISORDERLY CONDUCT 2 98 MISSION DISORDERLY CONDUCT 3 53 MISSION DISORDERLY CONDUCT 4 43 MISSION DISORDERLY CONDUCT 5 53 MISSION DISORDERLY CONDUCT 6 193 MISSION DISORDERLY CONDUCT 7 200 MISSION DISORDERLY CONDUCT 8 156 MISSION DISORDERLY CONDUCT 9 132 MISSION DISORDERLY CONDUCT 10 123 MISSION DISORDERLY CONDUCT 11 117 MISSION DISORDERLY CONDUCT 12 91 MISSION DISORDERLY CONDUCT 13 84 MISSION DISORDERLY CONDUCT 14 76 MISSION DISORDERLY CONDUCT 15 65 MISSION DISORDERLY CONDUCT 16 66 MISSION DISORDERLY CONDUCT 17 68 MISSION DISORDERLY CONDUCT 18 93 MISSION DISORDERLY CONDUCT 19 84 MISSION DISORDERLY CONDUCT 20 83 MISSION DISORDERLY CONDUCT 21 98 MISSION DISORDERLY CONDUCT 22 129 MISSION DISORDERLY CONDUCT 23 184 MISSION DRIVING UNDER THE INFLUENCE 0 115 MISSION DRIVING UNDER THE INFLUENCE 1 94 MISSION DRIVING UNDER THE INFLUENCE 2 80 MISSION DRIVING UNDER THE INFLUENCE 3 36 MISSION DRIVING UNDER THE INFLUENCE 4 16 MISSION DRIVING UNDER THE INFLUENCE 5 18 MISSION DRIVING UNDER THE INFLUENCE 6 10 MISSION DRIVING UNDER THE INFLUENCE 7 9 MISSION DRIVING UNDER THE INFLUENCE 8 6 MISSION DRIVING UNDER THE INFLUENCE 9 7 MISSION DRIVING UNDER THE INFLUENCE 10 5 MISSION DRIVING UNDER THE INFLUENCE 11 10 MISSION DRIVING UNDER THE INFLUENCE 12 11 MISSION DRIVING UNDER THE INFLUENCE 13 11 MISSION DRIVING UNDER THE INFLUENCE 14 13 MISSION DRIVING UNDE

In [31]:
#Finally, we pull out  the crime event data w.r.t category and time for Mission District.
crimeDistrictNorthern = spark.sql("SELECT PdDistrict, Category, hour(Time) As Hour, COUNT(*) AS Count FROM sf_crime GROUP BY 1,2,3 having PdDistrict = 'NORTHERN' ORDER BY 1, 2,3 ASC ")
display(crimeDistrictNorthern)

PdDistrict Category Hour Count NORTHERN ARSON 0 18 NORTHERN ARSON 1 27 NORTHERN ARSON 2 31 NORTHERN ARSON 3 38 NORTHERN ARSON 4 27 NORTHERN ARSON 5 14 NORTHERN ARSON 6 24 NORTHERN ARSON 7 11 NORTHERN ARSON 8 11 NORTHERN ARSON 9 7 NORTHERN ARSON 10 16 NORTHERN ARSON 11 6 NORTHERN ARSON 12 10 NORTHERN ARSON 13 7 NORTHERN ARSON 14 4 NORTHERN ARSON 15 13 NORTHERN ARSON 16 5 NORTHERN ARSON 17 15 NORTHERN ARSON 18 13 NORTHERN ARSON 19 15 NORTHERN ARSON 20 15 NORTHERN ARSON 21 15 NORTHERN ARSON 22 18 NORTHERN ARSON 23 26 NORTHERN ASSAULT 0 1209 NORTHERN ASSAULT 1 1188 NORTHERN ASSAULT 2 993 NORTHERN ASSAULT 3 513 NORTHERN ASSAULT 4 276 NORTHERN ASSAULT 5 231 NORTHERN ASSAULT 6 331 NORTHERN ASSAULT 7 483 NORTHERN ASSAULT 8 716 NORTHERN ASSAULT 9 774 NORTHERN ASSAULT 10 820 NORTHERN ASSAULT 11 946 NORTHERN ASSAULT 12 1199 NORTHERN ASSAULT 13 985 NORTHERN ASSAULT 14 980 NORTHERN ASSAULT 15 1157 NORTHERN ASSAULT 16 1068 NORTHERN ASSAULT 17 1011 NORTHERN ASSAULT 18 1015 NORTHERN ASSAULT 19 985 NORTHERN ASSAULT 20 961 NORTHERN ASSAULT 21 1098 NORTHERN ASSAULT 22 1107 NORTHERN ASSAULT 23 1012 NORTHERN BAD CHECKS 0 31 NORTHERN BAD CHECKS 1 1 NORTHERN BAD CHECKS 7 1 NORTHERN BAD CHECKS 8 5 NORTHERN BAD CHECKS 9 10 NORTHERN BAD CHECKS 10 11 NORTHERN BAD CHECKS 11 5 NORTHERN BAD CHECKS 12 30 NORTHERN BAD CHECKS 13 5 NORTHERN BAD CHECKS 14 4 NORTHERN BAD CHECKS 15 11 NORTHERN BAD CHECKS 16 9 NORTHERN BAD CHECKS 17 6 NORTHERN BAD CHECKS 18 3 NORTHERN BAD CHECKS 19 2 NORTHERN BAD CHECKS 20 2 NORTHERN BAD CHECKS 21 2 NORTHERN BAD CHECKS 23 1 NORTHERN BRIBERY 0 3 NORTHERN BRIBERY 1 3 NORTHERN BRIBERY 2 2 NORTHERN BRIBERY 4 1 NORTHERN BRIBERY 8 1 NORTHERN BRIBERY 9 2 NORTHERN BRIBERY 10 2 NORTHERN BRIBERY 11 4 NORTHERN BRIBERY 12 3 NORTHERN BRIBERY 13 1 NORTHERN BRIBERY 14 2 NORTHERN BRIBERY 15 2 NORTHERN BRIBERY 16 3 NORTHERN BRIBERY 17 5 NORTHERN BRIBERY 18 2 NORTHERN BRIBERY 19 6 NORTHERN BRIBERY 20 3 NORTHERN BRIBERY 21 3 NORTHERN BRIBERY 22 3 NORTHERN BRIBERY 23 3 NORTHERN BURGLARY 0 618 NORTHERN BURGLARY 1 404 NORTHERN BURGLARY 2 438 NORTHERN BURGLARY 3 441 NORTHERN BURGLARY 4 363 NORTHERN BURGLARY 5 366 NORTHERN BURGLARY 6 344 NORTHERN BURGLARY 7 488 NORTHERN BURGLARY 8 809 NORTHERN BURGLARY 9 623 NORTHERN BURGLARY 10 508 NORTHERN BURGLARY 11 465 NORTHERN BURGLARY 12 751 NORTHERN BURGLARY 13 450 NORTHERN BURGLARY 14 527 NORTHERN BURGLARY 15 623 NORTHERN BURGLARY 16 708 NORTHERN BURGLARY 17 884 NORTHERN BURGLARY 18 964 NORTHERN BURGLARY 19 752 NORTHERN BURGLARY 20 694 NORTHERN BURGLARY 21 730 NORTHERN BURGLARY 22 712 NORTHERN BURGLARY 23 659 NORTHERN DISORDERLY CONDUCT 0 77 NORTHERN DISORDERLY CONDUCT 1 73 NORTHERN DISORDERLY CONDUCT 2 91 NORTHERN DISORDERLY CONDUCT 3 69 NORTHERN DISORDERLY CONDUCT 4 48 NORTHERN DISORDERLY CONDUCT 5 73 NORTHERN DISORDERLY CONDUCT 6 95 NORTHERN DISORDERLY CONDUCT 7 92 NORTHERN DISORDERLY CONDUCT 8 73 NORTHERN DISORDERLY CONDUCT 9 72 NORTHERN DISORDERLY CONDUCT 10 43 NORTHERN DISORDERLY CONDUCT 11 27 NORTHERN DISORDERLY CONDUCT 12 32 NORTHERN DISORDERLY CONDUCT 13 33 NORTHERN DISORDERLY CONDUCT 14 34 NORTHERN DISORDERLY CONDUCT 15 19 NORTHERN DISORDERLY CONDUCT 16 25 NORTHERN DISORDERLY CONDUCT 17 21 NORTHERN DISORDERLY CONDUCT 18 13 NORTHERN DISORDERLY CONDUCT 19 21 NORTHERN DISORDERLY CONDUCT 20 20 NORTHERN DISORDERLY CONDUCT 21 12 NORTHERN DISORDERLY CONDUCT 22 27 NORTHERN DISORDERLY CONDUCT 23 45 NORTHERN DRIVING UNDER THE INFLUENCE 0 73 NORTHERN DRIVING UNDER THE INFLUENCE 1 81 NORTHERN DRIVING UNDER THE INFLUENCE 2 100 NORTHERN DRIVING UNDER THE INFLUENCE 3 40 NORTHERN DRIVING UNDER THE INFLUENCE 4 13 NORTHERN DRIVING UNDER THE INFLUENCE 5 2 NORTHERN DRIVING UNDER THE INFLUENCE 6 2 NORTHERN DRIVING UNDER THE INFLUENCE 7 5 NORTHERN DRIVING UNDER THE INFLUENCE 8 8 NORTHERN DRIVING UNDER THE INFLUENCE 9 6 NORTHERN DRIVING UNDER THE INFLUENCE 10 3 NORTHERN DRIVING UNDER THE INFLUENCE 11 3 NORTHERN DRIVING UNDER THE INFLUENCE 12 10 NORTHERN DRIVING UNDER THE INFLUENCE 13 4 NORTHERN DRIVING UNDER THE INFLUENCE 14

1. Based on above two-line charts, we observed that mission district and Northern district have similar pattern as Southern district with theft dominating the crime category. 
2. Based on the above analysis, theft dominates the crime events in these three districts. In order to decrease the number of crimes, police should put more resource to strike theft crime. 
3. Since the theft crime usually peak during 18 to 19 o'clock. Police department should be more cautious and allocate more resource to prevent the crime during these hours.

In [33]:
#Count the number of crimes in all the district and use a pie chart to help police alocate their resource.
crimeDistrict = spark.sql("SELECT PdDistrict, COUNT(*) AS Count FROM sf_crime GROUP BY 1 ORDER BY Count DESC")
display(crimeDistrict)

PdDistrict Count SOUTHERN 399785 MISSION 300076 NORTHERN 272713 CENTRAL 226255 BAYVIEW 221000 INGLESIDE 194180 TENDERLOIN 191746 TARAVAL 166971 PARK 125479 RICHMOND 116818 null 1

The above pie chart shows the percentage of the crime number of each district. The police resource should be ditributed accordingly.

#### (OLAP) Analysis of the resolutions of different catogaries

In [36]:
#find out the distribution of crimes in the resolution
temp = spark.sql("SELECT Resolution, count(*) FROM sf_crime GROUP BY Resolution ORDER BY Count(*) DESC")
display(temp)

Resolution count(1) NONE 1389500 ARREST, BOOKED 524979 ARREST, CITED 154789 LOCATED 34463 PSYCHOPATHIC CASE 29185 UNFOUNDED 23799 JUVENILE BOOKED 14158 COMPLAINANT REFUSES TO PROSECUTE 8089 DISTRICT ATTORNEY REFUSES TO PROSECUTE 7955 NOT PROSECUTED 7720 JUVENILE CITED 6587 PROSECUTED BY OUTSIDE AGENCY 5070 EXCEPTIONAL CLEARANCE 4258 JUVENILE ADMONISHED 3004 JUVENILE DIVERTED 694 CLEARED-CONTACT JUVENILE FOR MORE INFO 689 PROSECUTED FOR LESSER OFFENSE 85

In [37]:
%sql
 
SELECT category, 
        SUM(case when Resolution ='NONE' then 1 else 0 end)/count(*) as None,
        SUM(case when Resolution ='ARREST, BOOKED' then 1 else 0 end)/count(*) as ARREST_BOOKED,
        SUM(case when Resolution ='ARREST, CITED' then 1 else 0 end)/count(*) as ARREST_CITED,
        SUM(case when Resolution ='LOCATED' then 1 else 0 end)/count(*) as LOCATED,
        SUM(case when Resolution ='PSYCHOPATHIC CASE' then 1 else 0 end)/count(*) as PSYCHOPATHIC,
        SUM(case when Resolution ='UNFOUNDED' then 1 else 0 end)/count(*) as UNFOUNDED,
        SUM(case when Resolution ='JUVENILE BOOKED' then 1 else 0 end)/count(*) as JUVENILE_BOOKED,
        SUM(case when Resolution ='COMPLAINANT REFUSES TO PROSECUTE' then 1 else 0 end)/count(*) as COMPLAINANT_REFUSES_TO_PROSECUTE,
        SUM(case when Resolution ='DISTRICT ATTORNEY REFUSES TO PROSECUTE' then 1 else 0 end)/count(*) as DISTRICT_ATTORNEY_REFUSES_TO_PROSECUTE,
        SUM(case when Resolution ='NOT PROSECUTED' then 1 else 0 end)/count(*) as NO_PROSECUTED,
        SUM(case when Resolution ='JUVENILE CITED' then 1 else 0 end)/count(*) as JUVENILE_CITED,
        SUM(case when Resolution ='PROSECUTED BY OUTSIDE AGENCY' then 1 else 0 end)/count(*) as PROSECUTED_BY_OUTSIDE_AGENCY,
        SUM(case when Resolution ='EXCEPTIONAL CLEARANCE' then 1 else 0 end)/count(*) as EXCEPTIONAL_CLEARANCE,
        SUM(case when Resolution ='JUVENILE ADMONISHED' then 1 else 0 end)/count(*) as JUVENILE_ADMONISHED,
        SUM(case when Resolution ='JUVENILE DIVERTED' then 1 else 0 end)/count(*) as JUVENILE_DIVERTED,
        SUM(case when Resolution ='CLEARED-CONTACT JUVENILE FOR MORE INFO' then 1 else 0 end)/count(*) as CLEARED_CONTACT_JUVENILE_FOR_MORE_INFO,
        SUM(case when Resolution ='PROSECUTED FOR LESSER OFFENSE' then 1 else 0 end)/count(*) as PROSECUTED_FOR_LESSER_OFFENSE
FROM sf_crime 
GROUP BY category 
ORDER BY category DESC
                 


category None ARREST_BOOKED ARREST_CITED LOCATED PSYCHOPATHIC UNFOUNDED JUVENILE_BOOKED COMPLAINANT_REFUSES_TO_PROSECUTE DISTRICT_ATTORNEY_REFUSES_TO_PROSECUTE NO_PROSECUTED JUVENILE_CITED PROSECUTED_BY_OUTSIDE_AGENCY EXCEPTIONAL_CLEARANCE JUVENILE_ADMONISHED JUVENILE_DIVERTED CLEARED_CONTACT_JUVENILE_FOR_MORE_INFO PROSECUTED_FOR_LESSER_OFFENSE WEAPON LAWS 0.2730053071871908 0.5997121525591437 0.052037420167311325 3.148331384366286E-4 8.995232526760817E-4 0.007016281370873437 0.034316812089592516 0.002023927318521184 0.0042277592875775835 0.0029684267338310694 0.01178375461005667 6.746424395070612E-4 0.0017540703427183593 0.00859044706305658 3.148331384366286E-4 3.148331384366286E-4 4.4976162633804086E-5 WARRANTS 0.05407431519348189 0.9182572327602363 0.008729618560056817 3.3537517631856696E-4 3.55103127866718E-4 0.0013020448021779658 0.014766371733791021 2.0714349125558546E-4 3.55103127866718E-4 1.77551563933359E-4 2.56463370125963E-4 4.2415095828524644E-4 2.95919273222265E-4 2.9591927322226496E-5 2.465993943518875E-4 6.904783041852849E-5 1.1836770928890598E-4 VEHICLE THEFT 0.916099271733464 0.03999936809845026 0.0014612723337703985 1.5797538743463768E-4 7.898769371731884E-6 0.037827206521224 0.0018562108023569928 4.186347767017899E-4 5.529138560212319E-4 1.500766180629058E-4 9.715486327230218E-4 8.688646308905073E-5 2.7645692801061595E-4 3.949384685865942E-5 6.319015497385507E-5 2.369630811519565E-5 7.898769371731884E-6 VANDALISM 0.8779155429565997 0.07528929251501391 0.025599048759682577 2.1540768057625863E-4 0.001697412522940918 0.0037653262564730007 0.0025245780163537512 0.002300554028554442 0.0016457146796026159 2.9295444558371175E-4 0.006298520580049803 1.0339568667660414E-4 7.23769806736229E-4 0.0014216906918033069 1.723261444610069E-4 3.446522889220138E-5 0.0 TRESPASS 0.32325569438017376 0.39066275901074604 0.2598077021954856 5.141652527122217E-4 0.0026736593141035527 0.0034963237184431075 0.0037019898195279964 0.0028793254151884416 9.254974548819991E-4 0.0012339966065093321 0.006941230911614993 2.5708262635611086E-4 0.0011311635559668877 0.002210910586662553 1.5424957581366652E-4 1.0283305054244434E-4 5.141652527122217E-5 TREA 0.42857142857142855 0.42857142857142855 0.14285714285714285 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 0.0 SUSPICIOUS OCC 0.8824275272239073 0.03501814927154294 0.011150614091790563 0.0029337178658445625 0.00253592561284869 0.031984983342449404 0.0014917209487345234 0.007719655909701159 0.005046989209885138 0.005767987668440157 7.458604743672617E-4 0.0032196310476853463 0.003430958182089404 0.005208592312664711 4.59947292526478E-4 8.453085376162299E-4 1.2431007906121028E-5 SUICIDE 0.760061919504644 0.02631578947368421 0.0015479876160990713 0.014705882352941176 0.18188854489164086 0.00696594427244582 0.0 0.0 7.739938080495357E-4 0.0 0.0 0.0 0.00696594427244582 0.0 7.739938080495357E-4 0.0 0.0 STOLEN PROPERTY 0.12101589437389622 0.8151543183920612 0.013539651837524178 2.5229164914641325E-4 8.409721638213776E-5 0.0014296526784963418 0.02733159532419477 3.3638886552855103E-4 0.002270624842317719 0.004961735766546127 0.008662013287360189 0.0025229164914641325 0.0012614582457320662 7.568749474392397E-4 1.6819443276427552E-4 1.6819443276427552E-4 8.409721638213776E-5 SEX OFFENSES, NON FORCIBLE 0.4918793503480278 0.25754060324825984 0.004640371229698376 0.0069605568445475635 0.0 0.030162412993039442 0.02088167053364269 0.058004640371229696 0.09280742459396751 0.009280742459396751 0.0 0.01160092807424594 0.01160092807424594 0.004640371229698376 0.0 0.0 0.0 SEX OFFENSES, FORCIBLE 0.582950093680804 0.19323794924203713 0.019161982626469086 0.0015329586101175269 0.0012774655084312723 0.04292284108329075 0.009708737864077669 0.04104922500425822 0.06591722023505366 0.003832396525293817 0.0022994379151762903 0.006557656276613865 0.018140010219724067 0.003491739056378811 1.7032873445750298E-4 0.007749957417816386 0.0 SECONDARY CODES 0.5681158298168867 0.3092021214819403 0.008168479733653363 0.002709922

#### Note(Insight)
1. Based on the above chart, if we sorted the 'None', we will find that for the 'recovered vehicle', 'vehicle theft' and 'larceny/theft', more than 90 percent cases are marked 'None'. That means more than 90 percent of cases have no resolution at all. In order to improve the resolution of the crime related to theft, we recommend the police increase surveillance effort. 
2. From the previous analysis, beside theft, drug related crimes account for a large proportion of the total crime in SF. In this resolution distribution chart, 81 percent of drug crime were resolved by arrest and booked. That means we have a good record of the person who was suspect or prosecuted by drug related charge before on file. In order to decrease the future crime rate in this category, police can focuses on the individuals and groups who drive most of that activity, especially those with the criminal records.

### Clustering

In [40]:
#I want to use the location of crime to do the clustering. So I first pull out the latitude and longitude infromation from the original dataframe.
df_cluster =spark.sql("SELECT IncidntNum as id, X, Y FROM sf_crime")
display(df_cluster)

id X Y 146196161 -122.403404791479 37.775420706711 150045675 -122.48560378101 37.7388214326705 140632022 -122.396535107224 37.7106603302503 150383259 -122.400130573297 37.7300925390327 040753980 -120.5 90 040855122 -122.386667033903 37.7898821569191 066085191 -122.38750147945 37.716878646429 050908404 -120.5 90 090768064 -122.399686082806 37.739901780585 111027676 -122.412933062384 37.7739274524819 060508925 -122.419834938618 37.7863580362743 110318854 -122.398951090192 37.7338473162583 131007921 -122.435709983124 37.7443688410754 090333811 -122.428006990942 37.7724726132375 050951798 -120.5 90 031057056 -122.386667033903 37.7898821569191 061079105 -122.42831052169 37.7823629069329 050326981 -122.395251009931 37.7553443519568 120444392 -122.458724812805 37.7802795296617 051130171 -120.5 90 120767417 -122.450482980088 37.708170291703 180210729 -122.42485988902867 37.78619644725339 160452270 -122.47813089080685 37.71940181416355 150103687 -122.411123069982 37.7363322364486 150105809 -122.436910088417 37.7998528145322 150098759 -122.410541166987 37.7709130566165 150109704 -122.397744427103 37.7299346936044 150106998 -122.4094512966 37.7512276969557 150105116 -122.380348349699 37.7288276251864 150109845 -122.393859499886 37.7779770738329 150103502 -122.470595847264 37.7488679961125 150106885 -122.415434713819 37.7939455886301 150101346 -122.499264595602 37.7410543982056 150109964 -122.404718428538 37.797221829197 150099246 -122.406481972743 37.7827317884887 150108518 -122.415885065788 37.7835160564153 150108342 -122.416583456133 37.7786591541508 150104726 -122.412784096502 37.7848656939526 150099014 -122.424060357638 37.7980723406312 150102350 -122.419520367886 37.764228935718 150107708 -122.412970537591 37.7857883766888 150101255 -122.412596970637 37.7811192121542 150109083 -122.407612885387 37.7884486019705 150105473 -122.453994369282 37.7164758973784 150106001 -122.43307534135 37.726854596498 150101114 -122.412930522059 37.7838344374141 150103502 -122.470595847264 37.7488679961125 150106948 -122.401548064995 37.7929509214078 150100360 -122.410135236377 37.7841889075267 150105809 -122.436910088417 37.7998528145322 150109641 -122.396320819026 37.7372763126101 150102639 -122.400302421513 37.7777992290419 150107463 -122.41565370001 37.7787196511256 150106352 -122.410962664755 37.752684919614 150102253 -122.410774214375 37.8084350175869 150100166 -122.411162307172 37.719668771199 150106186 -122.411800134753 37.7840890366868 150060275 -122.42158168137 37.7617007179518 150108449 -122.402224095303 37.781960649289 150104823 -122.440660894634 37.7205097496756 150102877 -122.37934537972 37.7357419068572 150103831 -122.420303482492 37.770896885856 150109497 -122.436749760121 37.7219915768943 150098925 -122.487429256289 37.7800894126441 150107786 -122.426390080828 37.7185612508135 150102366 -122.403143544447 37.7390301515332 150101277 -122.43340738035 37.7992089703227 150108637 -122.404767635853 37.7946646538154 150101669 -122.409853729941 37.7833862379382 150106506 -122.447517523122 37.7810694335442 150099531 -122.41029351969 37.7822305870036 150105683 -122.373456171525 37.7296796672727 150108756 -122.407729769991 37.7931773025837 150102667 -122.409660417788 37.7857429288295 150108239 -122.381400971393 37.7294904424423 150103110 -122.435141194073 37.7330256190757 150104754 -122.414743835382 37.778719262789 150106700 -122.403750443804 37.7229277838502 150099105 -122.444707063455 37.7249307267936 150104011 -122.411071423064 37.7817511307229 150104362 -122.394698550916 37.7645975666681 150109431 -122.481003317121 37.7803821307757 150100031 -122.403404791479 37.775420706711 150100586 -122.439992498876 37.7171486722748 150107316 -122.393324841737 37.7188264259962 150106619 -122.469357481945 37.7779050116432 150100871 -122.410157073911 37.7842973843312 150104544 -122.45275843965 37.7654433196318 150107388 -122.402402526411 37.7460992600025 150098232 -122.451781767894 37.7870853907529 150105655 -122.387067977547 37.7554460266182 150108687 -122.42

In [41]:
#Becuase the input of kmeans package should be feature vectors with float type. We need to transform the data frame.
FEATURES_COL = ['X', 'Y']
#df_cluster = df_cluster.select(*(df_cluster[c].cast("float").alias(c) for c in df_cluster.columns[1:]))
#df_cluster.show()
for col in df_cluster.columns:
    if col in FEATURES_COL:
        df_cluster = df_cluster.withColumn(col,df_cluster[col].cast('float'))
df_cluster.show()



+---------+-----------+---------+
 id| X| Y|
+---------+-----------+---------+
146196161|-122.403404| 37.77542|
150045675| -122.4856|37.738823|
140632022| -122.39654| 37.71066|
150383259| -122.40013| 37.73009|
040753980| -120.5| 90.0|
040855122|-122.386665|37.789883|
066085191|-122.387505|37.716877|
050908404| -120.5| 90.0|
090768064| -122.39969|37.739902|
111027676| -122.41293|37.773926|
060508925| -122.41984|37.786358|
110318854| -122.39895| 37.73385|
131007921| -122.43571| 37.74437|
090333811| -122.42801|37.772472|
050951798| -120.5| 90.0|
031057056|-122.386665|37.789883|
061079105|-122.428314|37.782364|
050326981| -122.39525|37.755344|
120444392|-122.458725| 37.78028|
051130171| -120.5| 90.0|
+---------+-----------+---------+
only showing top 20 rows

In [42]:
#Transform data fram into vector sets
from pyspark.ml.feature import VectorAssembler

vecAssembler = VectorAssembler(inputCols=FEATURES_COL, outputCol="features")
df_kmeans = vecAssembler.transform(df_cluster).select('id', 'features')
df_kmeans.show()

+---------+--------------------+
 id| features|
+---------+--------------------+
146196161|[-122.40340423583...|
150045675|[-122.48560333251...|
140632022|[-122.39653778076...|
150383259|[-122.40013122558...|
040753980| [-120.5,90.0]|
040855122|[-122.38666534423...|
066085191|[-122.38750457763...|
050908404| [-120.5,90.0]|
090768064|[-122.39968872070...|
111027676|[-122.41293334960...|
060508925|[-122.41983795166...|
110318854|[-122.39894866943...|
131007921|[-122.43570709228...|
090333811|[-122.42800903320...|
050951798| [-120.5,90.0]|
031057056|[-122.38666534423...|
061079105|[-122.42831420898...|
050326981|[-122.39524841308...|
120444392|[-122.45872497558...|
051130171| [-120.5,90.0]|
+---------+--------------------+
only showing top 20 rows

In [43]:

from pyspark.ml.clustering import KMeans
from pyspark.ml.evaluation import ClusteringEvaluator



# Trains a k-means model.
kmeans = KMeans().setK(5).setSeed(1)
model = kmeans.fit(df_kmeans)

# Make predictions
predictions = model.transform(df_kmeans)

# Evaluate clustering by computing Silhouette score
evaluator = ClusteringEvaluator()

silhouette = evaluator.evaluate(predictions)
print("Silhouette with squared euclidean distance = " + str(silhouette))

# Shows the result.
centers = model.clusterCenters()
print("Cluster Centers: ")
for center in centers:
    print(center)




Silhouette with squared euclidean distance = 0.5597636034987211
Cluster Centers: 
[-122.47652315 37.74754932]
[-120.5 90. ]
[-122.40918075 37.73509192]
[-122.40945435 37.78490068]
[-122.43321491 37.77299508]

In [44]:
predictions.show()

+---------+--------------------+----------+
 id| features|prediction|
+---------+--------------------+----------+
146196161|[-122.40340423583...| 3|
150045675|[-122.48560333251...| 0|
140632022|[-122.39653778076...| 2|
150383259|[-122.40013122558...| 2|
040753980| [-120.5,90.0]| 1|
040855122|[-122.38666534423...| 3|
066085191|[-122.38750457763...| 2|
050908404| [-120.5,90.0]| 1|
090768064|[-122.39968872070...| 2|
111027676|[-122.41293334960...| 3|
060508925|[-122.41983795166...| 3|
110318854|[-122.39894866943...| 2|
131007921|[-122.43570709228...| 2|
090333811|[-122.42800903320...| 4|
050951798| [-120.5,90.0]| 1|
031057056|[-122.38666534423...| 3|
061079105|[-122.42831420898...| 4|
050326981|[-122.39524841308...| 2|
120444392|[-122.45872497558...| 4|
051130171| [-120.5,90.0]| 1|
+---------+--------------------+----------+
only showing top 20 rows

In [45]:
%matplotlib inline
import matplotlib.pyplot as plt
import seaborn as sns; sns.set()  
import numpy as np

In [46]:
pd_cluster = df_cluster.select("*").toPandas()
pd_predictions = predictions.select("*").toPandas()

/databricks/spark/python/pyspark/sql/dataframe.py:2165: UserWarning: toPandas attempted Arrow optimization because 'spark.sql.execution.arrow.enabled' is set to true; however, failed by the reason below:
 Unsupported type in conversion to Arrow: VectorUDT
Attempting non-optimization as 'spark.sql.execution.arrow.fallback.enabled' is set to true.
 warnings.warn(msg)

In [47]:
df_cluster.createOrReplaceTempView("cluster")
update_cluster = spark.sql("SELECT COUNT(*) FROM cluster where Y=90.0")
display(update_cluster)
                          

count(1) 143

Note:We found there are 143 points are located very far away from all the other points. For better visualization effect, we did not include them in the scatter plot below. Therefore, only four centers and colors are shown on the plot.

In [49]:
x_center = []
y_center = []

for x, y in centers:
  x_center.append(x)
  y_center.append(y)

print(x_center)
print(y_center)

[-122.47652314917055, -120.5, -122.40918074698594, -122.40945434545321, -122.43321490801564]
[37.747549317198335, 90.0, 37.7350919222457, 37.784900678524465, 37.77299507855718]

In [50]:
plt.scatter(pd_cluster["X"], pd_cluster["Y"],c=pd_predictions["prediction"], s=0.5, cmap='viridis')
plt.axis([-122.52, -122.36, 37.7, 37.85])
plt.scatter(x_center, y_center, c='black', s=100, alpha=0.5);

#### Note(Insight)
Based on the above K means model, we located five central point of crime events. These peice information can be used to alocate police resource. For example, in the daily peak of crime, gorverment can deploy more police to patrol around these loacations to prevent crime.

### Conclusion. 

Point 1:  San Francisco as one of the largest cities on the West Coast, is the cultural, commercial, and financial center. Everyday, millions of arties, businessman and travelers come to the city. The most import question in their mind is whether San Francisco is safe or not. My analysis is to better understand the crime pattern in San Francisco so that we can have a more constructive advice to our travelers and to our policy makers.  
Point 2:  To accomplish this goal, we used the Spark SQL and Spark Data frame to analysis the data. First we use data visualization and some data manipulation to explore the patterns from different prospects. Then we used k means clustering to study the data.  
Point 3:  The data has obvious yearly and daily patterns. And the crime tends to cluster in certain area. We suggest travelers to travels in a lower crime rate time and avoid the dangerous area. And police can plan the deployment based on the grime trends during recent years and based on the distribution of the different crimes.